docs/experiments/extract-mode.md

```diff
@@ -0,0 +1,103 @@
+---
+id: extract-mode
+append_help_link: true
+description: "This article explains the extract mode, which allows for easier handling of files containing more than one language."
+---
+
+# Extract mode
+
+## Introduction
+
+Extract mode enables you to run existing rules on subsections of files where the rule language is different than the language of the file. For example, running a Javascript rule on code contained inside of script tags in an HTML document.
+
+## Example of extract mode
+
+Without extract mode, writing rules to validate template, Markdown or configuration files which contain code in another language can be burdensome and require significant rule duplication.
+
+Let's take the following Bash rule as an example (a simplified version of [our `curl-eval` rule](https://github.com/returntocorp/semgrep-rules/blob/release/bash/curl/security/curl-eval.yaml)):
+
+```yaml
+rules:
+  - id: curl-eval
+    severity: WARNING
+    languages:
+      - bash
+    message: Evaluating data from a `curl` command is unsafe.
+    mode: taint
+    pattern-sources:
+      - pattern: |
+          $(curl ...)
+      - pattern: |
+          `curl ...`
+    pattern-sinks:
+      - pattern: eval ...
+```
+
+Usually, this rule would only be run against Bash files. However, our project might contain Dockerfiles or Python scripts which invoke Bash commands&mdash;without an extract mode rule Semgrep won't know about any contained Bash, and therefore, will **not** run any Bash rules against those contained commands.
+
+However, with extract mode we can provide Semgrep with instructions on how to extract any Bash commands we might be using in a Docker `RUN` instruction or as an argument to Python's `os.system` standard library function.
+
+```yaml
+rules:
+  - id: extract-docker-run-to-bash
+    mode: extract
+    languages:
+      - dockerfile
+    pattern: RUN $...CMD
+    extract: $...CMD
+    dest-language: bash
+  - id: extract-python-os-system-to-bash
+    mode: extract
+    languages:
+      - python
+    pattern: os.system("$CMD")
+    extract: $CMD
+    dest-language: bash
+```
+
+By adding the extract mode rules show above, Semgrep can match the Bash contained in the below Python file, and will report the contained Bash as  matching against the `curl-eval` rule.
```

Comment:
Blocking:
- Do not indicate directions in this manner in one sentence. In this particular case, code snippets below the paragraph are the following code snippets, code snippets that are above the paragraph can be indicated as: in the previous code snippet.
-  Avoid will when possible. Documentation needs to be written mostly in the present simple tense: See [will](https://developers.google.com/style/word-list?hl=en#will) in the Google developer style guide. 

```suggestion
By adding the extract mode rules as shown in the previous code snippet, Semgrep matches the Bash contained in the following Python file and reports the contained Bash as matching against the `curl-eval` rule.
```